llext: Fix off-by-one in RISC-V truncation check

WorldofJARcraft opened the pull request:
The RISC-V architecture-specific relocations need to check whether each required relocation can fit into the modified instruction's immediate. All immediates in RISC-V are encoded as two's complement. The current truncation check has an off-by-one error for checking the maximum negative distance, as two's complement encoding can represent a negative value that is the maximum positive value plus one, causing LLEXT to refuse loading valid code.
This commit adds an additional condition to the check that fixes the aforementioned issue.

For a concrete example of the problem, consider the CB-type instruction: It has an 8-bit immediate with an explicit LSB of 0. The current code accepts relocations for CB-type instruction within a range of [-255,+255] from the operation's location, relying on standard RISC-V alignment requirements to ensure that the LSB is 0.
However, while the maximum positive jump distance of 255 is correct, the maximum negative jump distance for a CB-type instruction is actually -256 bit. Here is an example in a RISC-V assembler for a jump with -256 byte offset: https://luplab.gitlab.io/rvcodecjs/#q=c.beqz+a0,+-256&abi=false&isa=AUTO
Here is an example for trying to jump +256 byte (interpreted as -256 byte jump): https://luplab.gitlab.io/rvcodecjs/#q=c.beqz+a0,+256&abi=false&isa=AUTO
The RISC-V ISA manual states (somewhat misleadingly in my opinion) that the CB-type instruction can perform jumps in a [-256,+256] byte range: https://github.com/riscv/riscv-isa-manual/releases/download/riscv-isa-release-a8395e7-2024-12-16/riscv-privileged.pdf, page 167; I have raised a separate issue for the ISA manual: https://github.com/riscv/riscv-isa-manual/issues/1781

WorldofJARcraft committed:
llext: Fix off-by-one in RISC-V truncation check

The RISC-V architecture-specific relocations need to check whether
each required relocation can fit into the modified instruction's
immediate. All immediates in RISC-V are encoded as two's complement.
The current truncation check has an off-by-one error for checking
the maximum negative distance, as two's complement encoding can
represent a negative value that is the maximum positive value plus
one, causing LLEXT to refuse loading valid code.
This commit adds an additional condition to the check that fixes
the aforementioned issue.

Signed-off-by: Eric Ackermann <eric.ackermann@cispa.de>

teburd reviewed (dismissed):
The change looks great, it would be a good value add here to have a test validating this edge that clearly got tripped on

WorldofJARcraft committed:
tests: llext: Add RISC-V CB-Type edge case test

All immediates in RISC-V are encoded as two's complement. This commit
adds a test for relocating jumps that utilize the full range of the
immediate, in both positive and negative direction.
To this end, the test uses the compressed b-type (CB) instruction to
branch to its maximum negative (-256) and maximum positive (+254)
targets.
In case of test failure, expect relocating the corresponding llext to
fail.

Signed-off-by: Eric Ackermann <eric.ackermann@cispa.de>

WorldofJARcraft commented:
@teburd a test that uses both the maximum positive and negative values of the immediate has been added in a separate commit.
I have tested manually that RISC-V tests succeed when the first commit in this PR has been applied and fail when it has not been applied. The error message is the one I was expecting: the llext could not be relocated.

teburd commented:
Small formatting issue, but the tests are *much* appreciated by me

teburd reviewed (dismissed):
Approved, though I expect a small force push will cause some churn again. Happy to reapprove again if needed

WorldofJARcraft commented:
This branch has been rebased to main; I have tested again that undoing the fix in the first commit causes tests to fail in the expected manner.

WorldofJARcraft commented on tests/subsys/llext/src/riscv_edge_case_cb_type_trigger.S:
```diff
@@ -0,0 +1,64 @@
+
+    /*
+     * 2+4 bytes for _backward_jump_target itself
+     * need to insert 122 padding bytes
```
You are correct, the comment is out-of-date, it should state that 252 bytes of padding are needed. The `c.addi` and `ret` instructions are 2 bytes each, and we need to have exactly 256 bytes between `_do_jump` and `_backward_jump_target` to trigger the edge case.
I will update this and also change `ret` to a `c.jr` compressed jump to make this more obvious.

pillo79 commented on tests/subsys/llext/src/riscv_edge_case_cb_type_trigger.S:
```diff
@@ -0,0 +1,92 @@
+     * we pad with 0xff here, which at the time of writing is an invalid opcode and causes
+     * an exception should it ever be executed
+     */
+    .fill 126, 2, 0xff
```
This will actually fill the file with sequences of `0x00 0xff`. Is this intentional?

pillo79 reviewed (dismissed):
One minor clarification - otherwise LGTM!

pillo79 reviewed (approved):
Love the detail work, thanks for this!
